src/HelperPluginManager.php

```diff
@@ -22,59 +23,117 @@
 {
-    /**
-     * Default set of helpers factories
-     *
-     * @var array
-     */
-    protected $factories = [
-        'flashmessenger' => 'Zend\View\Helper\Service\FlashMessengerFactory',
-        'identity'       => 'Zend\View\Helper\Service\IdentityFactory',
-    ];
-
-    /**
-     * Default set of helpers
-     *
-     * @var array
-     */
-    protected $invokableClasses = [
-        // basepath, doctype, and url are set up as factories in the ViewHelperManagerFactory.
-        // basepath and url are not very useful without their factories, however the doctype
-        // helper works fine as an invokable. The factory for doctype simply checks for the
-        // config value from the merged config.
-        'basepath'            => 'Zend\View\Helper\BasePath',
-        'cycle'               => 'Zend\View\Helper\Cycle',
-        'declarevars'         => 'Zend\View\Helper\DeclareVars',
-        'doctype'             => 'Zend\View\Helper\Doctype', // overridden by a factory in ViewHelperManagerFactory
-        'escapehtml'          => 'Zend\View\Helper\EscapeHtml',
-        'escapehtmlattr'      => 'Zend\View\Helper\EscapeHtmlAttr',
-        'escapejs'            => 'Zend\View\Helper\EscapeJs',
-        'escapecss'           => 'Zend\View\Helper\EscapeCss',
-        'escapeurl'           => 'Zend\View\Helper\EscapeUrl',
-        'gravatar'            => 'Zend\View\Helper\Gravatar',
-        'htmltag'             => 'Zend\View\Helper\HtmlTag',
-        'headlink'            => 'Zend\View\Helper\HeadLink',
-        'headmeta'            => 'Zend\View\Helper\HeadMeta',
-        'headscript'          => 'Zend\View\Helper\HeadScript',
-        'headstyle'           => 'Zend\View\Helper\HeadStyle',
-        'headtitle'           => 'Zend\View\Helper\HeadTitle',
-        'htmlflash'           => 'Zend\View\Helper\HtmlFlash',
-        'htmllist'            => 'Zend\View\Helper\HtmlList',
-        'htmlobject'          => 'Zend\View\Helper\HtmlObject',
-        'htmlpage'            => 'Zend\View\Helper\HtmlPage',
-        'htmlquicktime'       => 'Zend\View\Helper\HtmlQuicktime',
-        'inlinescript'        => 'Zend\View\Helper\InlineScript',
-        'json'                => 'Zend\View\Helper\Json',
-        'layout'              => 'Zend\View\Helper\Layout',
-        'paginationcontrol'   => 'Zend\View\Helper\PaginationControl',
-        'partialloop'         => 'Zend\View\Helper\PartialLoop',
-        'partial'             => 'Zend\View\Helper\Partial',
-        'placeholder'         => 'Zend\View\Helper\Placeholder',
-        'renderchildmodel'    => 'Zend\View\Helper\RenderChildModel',
-        'rendertoplaceholder' => 'Zend\View\Helper\RenderToPlaceholder',
-        'serverurl'           => 'Zend\View\Helper\ServerUrl',
-        'url'                 => 'Zend\View\Helper\Url',
-        'viewmodel'           => 'Zend\View\Helper\ViewModel',
+    protected $instanceOf = Helper\HelperInterface::class;
+
+    protected $config = [
```

Comment:
There's a method behind the madness.

You can define keys up-front if you will for a number of the services. However, one of them in particular is not supported: invokables. The reason is because internally, `configure()` takes the `invokables` key and creates aliases and factories (pointing a class to the `InvokableFactory`) out of them. As such, these, in particular, need to be merged with incoming configuration and passed to the constructor *anyways*.

Additionally, there's another reason: Internally, the `Navigation\PluginManager` implementation extends `HelperPluginManager`. This is done to automate injection of the renderer and/or translator into the resolved helper on instantiation. However, that plugin manager *does not* and should not use the same configuration as the view helper plugin manager, as it should only resolve the various navigation helpers. As such, having these as a single array makes the inheritance story far easier.